### PR TITLE
App service accepts archives

### PR DESCRIPTION
**Pull request by cantino**
Allow a `.tgz` file to be provided instead of a directory to the app service's register mutation. The binary and manifest must be in the root of the tarball (not in a nested directory), and detection is in place to warn if the user makes a mistake.

I've tested in the SDK and on my Mac. I'd appreciate if someone could test on hardware.

### CODE REVIEW

File: services/app-service/src/registry.rs [hunk 1]
```diff
@@ -219,18 +224,81 @@ impl AppRegistry {
             return Err(AppError::RegisterError {
                 err: format!("{} does not exist", path),
             });
-        }
+        } else if !app_path.is_dir() {
+            match app_path.extension().and_then(OsStr::to_str) {
```
Comment (cantino) [inline services/app-service/src/registry.rs]: If I were to move some of this into a separate method or file, where would you suggest it should go?

**Comment by jacoffey3** [comment]
Docs!

**Comment by cantino** [comment]
Docs added.

File: docs/tutorials/app-register.rst [hunk 2]
```diff
@@ -108,7 +108,8 @@ It has the following schema::
         }
      }
      
-The ``path`` input parameter specifies the directory where the application and manifest files reside.
+The ``path`` input parameter specifies the path to either a directory where the application and manifest files reside,
+or a the path to a ``.tgz`` archive file containing the application and manifest in its root.
```
Comment (plauche) [inline docs/tutorials/app-register.rst]: nit: `or the path`

File: services/app-service/src/registry.rs (hunk 1 above)
Comment (plauche) [reply services/app-service/src/registry.rs]: Yea this function is becoming a monster. You could split out a few private functions in this file. I potentially see an `extract_archive` from 230:279 and `validate_application` from 223:328.

File: docs/tutorials/app-register.rst (hunk 2 above)
Comment (cantino) [reply docs/tutorials/app-register.rst]: That's not a nit, that's egregious! Thanks. :)

File: services/app-service/src/registry.rs (hunk 1 above)
Comment (cantino) [reply services/app-service/src/registry.rs]: I did `extract_archive` but not `validate_application` for now. I have `extract_archive` doing implicit returns, which I think is readable enough.

**Pull request merged**